cleanup package `registered` and `install`

caesarxuchao opened the pull request:
fix #18531 

Some highlights of the PR:
1. in package `registered`, I remove the hard-coded list of valid versions and the hard-coded list of versions that are supported by default
2. The `install` package in each group initiate the group installation process. It depends on the `registered` package to check if a version should be enabled and keep records of enabled groups. The `init()` function in the `register` package in each version is changed to an exported function named `Register()`, and `install` will call it to add the version to the `Scheme`.
3. The `import_known_versions.go` in each binary imports the `install` package and calls a function exported by the `registered` package to check if all versions specified in KUBE_API_VERSIONS are installed.
4. [edit] The order of versions in KUBE_API_VERSIONS is not significant anymore. The preference of versions is specified in `availableVersions` in each `install` package (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/install/install.go#L42).

~~Note this PR doesn't implement exactly what I wrote in #18531. Specifically this PR doesn't distinguish `registered versions` from `enabled versions`, because I find it's useless to have versions that are registered but not enabled.~~

cc @nikhiljindal

Commit:
Refactor the API registration and installation

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,97 @@ import (
+// IsEnvReuqestedVersion returns if the version is requested via the
+// KUBE_API_VERSIONS environment variable. If the environment variable is empty,
+// then it always returns true.
+func IsEnvReuqestedVersion(v unversioned.GroupVersion) bool {
```
`IsEnvExcludedVersion` to make the "no env set" case be more expected?

caesarxuchao replied:
But I'm afraid that will lead people to think the versions listed in the Env are the excluded versions. I think explain the "no env set" case in the comment is clearer.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,97 @@ import (
+var (
+	// registeredVersions represents all enabled API versions. Please call
+	// RegisterVersions() to register the versions.
+	registeredVersions = map[unversioned.GroupVersion]struct{}{}
```
In the API Machinery call, we discussed splitting into `registeredVersion`, which is the list of all available group/versions and `enabledVersions`, which seems to be how this is godoced.

Can you make that split here in the variables even if you just update the `enabledVersions` list at the same time.  I think it will help us reason about the stages.

caesarxuchao replied:
@deads2k sorry about the godoc, I should have updated it after a grep&sed.

I'm thinking probably we don't need the `registeredVersion` we talked about in the call. We used to need it because we compare KUBE_API_VERSIONS with `registeredVersion` to make sure the versions in the env variable makes sense, but we have another mechanism to check that now.

What do you think? Why do we need to keep record of the versions that are compiled but not enabled?

caesarxuchao replied:
Note in the last commit, I ran a grep/sed to `s/enabled/registered`, so logically all `registeredXXX` variables in this PR represent `enabledXXX`. And there is no variable represent the concept of registered versions in this PR.

Sorry for the confusion.

deads2k replied:
> What do you think? Why do we need to keep record of the versions that are compiled but not enabled?

I think we might want to know all the available versions even when a subset is chosen.  It would allow us to do things like look at a whitelist (env var), compare against the available, and tell them which versions are disabled.  There might be other ways to do it, but I think the approach you've laid out in this pull of unconditional registration followed by inspection of enablement will make it easier to reason about the code and debug it after the fact.

caesarxuchao replied:
I see. I'll add back the `registeredVersiosns` list.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,97 @@ import (
+	// envRequestedVersions represents the versions requested via the KUBE_API_VERSIONS
+	// environment variable. The install package of each group checks this list before
+	// add their versions to the latest package and Scheme.
+	envRequestedVersions = map[unversioned.GroupVersion]struct{}{}
```
So in keeping with `registered` and `enabled`, this would not affect the `registered` list, but it would prevent versions from becoming `enabled`, right?

caesarxuchao replied:
Yes.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,97 @@ import (
+
+// ValidateEnvRequestedVersionsOrDie panics if KUBE_API_VERSIONS contains
+// versions that are not enabled.
+func ValidateEnvRequestedVersionsOrDie() {
```
I'd prefer to leave the dying to calling the code.  I wouldn't expect to see the call in many spots.

caesarxuchao replied:
Done.

deads2k commented on pkg/api/install/install.go:
```diff
@@ -128,3 +130,15 @@ func interfacesFor(version unversioned.GroupVersion) (*meta.VersionInterfaces, e
+	for _, v := range externalVersions {
+		switch v {
+		case v1.SchemeGroupVersion:
+			v1.Register()
```
Eventually this should be gated on a `register.IsEnabled(GroupVersion)` call, right?

```go
switch v {
case v1.SchemeGroupVersion && register.IsEnabled(v):
    v1.Register()
```

caesarxuchao replied:
I added the guard code.

deads2k commented:
I like this direction a lot.  I think making the distinction between `enabledVersions` and `registeredVersions` here will help a lot since you're ripping up those methods now.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,126 @@ import (
+// IsEnvReuqestedVersion returns if the version is requested via the
+// KUBE_API_VERSIONS environment variable. If the environment variable is empty,
+// then it always returns true.
+func IsEnvReuqestedVersion(v unversioned.GroupVersion) bool {
```
> But I'm afraid that will lead people to think the versions listed in the Env are the excluded versions. I think explain the "no env set" case in the comment is clearer.

Ok, how about `IsAllowedVersion`?  Also, the name is currently typoed: Requested

caesarxuchao replied:
I'll change to `IsAllowedVersion`. Thanks.

caesarxuchao replied:
Done.

caesarxuchao commented on docs/admin/kube-apiserver.md:
```diff
@@ -99,14 +99,14 @@ kube-apiserver
       --ssh-keyfile="": If non-empty, use secure SSH proxy to the nodes, using this user keyfile
       --ssh-user="": If non-empty, use secure SSH proxy to the nodes, using this user name
-      --storage-versions="componentconfig/v1alpha1,extensions/v1beta1,v1": The versions to store resources with. Different groups may be stored in different versions. Specified in the format "group1/version1,group2/version2...". This flag expects a complete list of storage versions of ALL groups registered in the server. It defaults to a list of preferred versions of all registered groups, which  ...
+      --storage-versions="componentconfig/v1alpha1,extensions/v1beta1,metrics/v1alpha1,v1": The versions to store resources with. Different groups may be stored in different versions. Specified in the format "group1/version1,group2/version2...". This flag expects a complete list of storage versions of ALL groups registered in the server. It defaults to a list of preferred versions of all register ...
```
This addition is caused by the apiserver's dependency on the client, so it will install all the versions listed in `pkg/client/unversioned/import_known_versions.sh`. 

I'm not sure why the `metrics` group isn't there before. Having it in the list looks correct anyway.

deads2k commented on pkg/api/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
 		return
 	}
 
+	registered.RegisterVersions(availableVersions...)
```
This should be done first.  You always want to register even if you have no external versions enabled.  In this order, you might return before registering.

caesarxuchao replied:
Done.

deads2k commented on pkg/api/install/install.go:
```diff
@@ -128,3 +135,19 @@ func interfacesFor(version unversioned.GroupVersion) (*meta.VersionInterfaces, e
+	// add the enabled external versions to Scheme
+	for _, v := range externalVersions {
+		if !registered.IsEnabledVersion(v) {
+			glog.V(4).Infof("Version %s is not enabled, so it will not be added to the Scheme.", v)
```
If we got here, something went wrong.  Let's log this at an `Errorf`

caesarxuchao replied:
Done.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,112 @@ import (
+		enabledVersions[v] = struct{}{}
+	}
+	if len(unregisteredVersions) != 0 {
+		fmt.Errorf("Please register versions before enabling them: %v", unregisteredVersions)
```
`return fmt.Errorf(...`

caesarxuchao replied:
Done.

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,112 @@ import (
-	ret := []unversioned.GroupVersion{}
-	for _, v := range RegisteredGroupVersions {
+// EnabledVersionsForGroup returns all enabled versions for a group.
+func EnabledVersionsForGroup(group string) (ret []unversioned.GroupVersion) {
```
Add a `RegisteredVersionsForGroup` for congruence?

deads2k commented on pkg/api/registered/registered.go:
```diff
@@ -25,71 +26,112 @@ import (
-			return true
-		}
+// IsEnabledVersion returns if a version is enabled.
+func IsEnabledVersion(v unversioned.GroupVersion) bool {
```
Add an `IsRegisteredVersion` for congruence and for use in `EnableVersions`?

deads2k commented on pkg/apis/componentconfig/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
 		return
 	}
 
+	registered.RegisterVersions(availableVersions...)
```
congruent comments to the other `install.go` here and for the remainder.

deads2k commented:
A few relatively minor comments.  Go ahead and squash while addressing them.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
 		return
 	}
 
+	registered.RegisterVersions(availableVersions...)
```
I will suggest moving this to be the first line of init() method.
The compute externalVersions and enable them.
That, I think, makes the code more readable.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
 
+	registered.RegisterVersions(availableVersions...)
+
+	if err := registered.EnableVersions(externalVersions...); err != nil {
```
EnableVersion just marks it as enabled. The actual enabling is done by addVersionsToScheme() method below, right?
How about renaming EnableVersion to AddEnabledVersion()?

Also, I will suggest calling addVersionsToScheme() first and then call EnableVersion for readability.

caesarxuchao replied:
> EnableVersion just marks it as enabled. The actual enabling is done by addVersionsToScheme() method below, right? How about renaming EnableVersion to AddEnabledVersion()?

That's what I named the function at first, but then I think AddEnabledVersion() can be interpreted as "enable this version" so I go for the short name. Anyway, since you also think AddEnabledVersion is better, I'll change it back. Glad our thoughts align.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -43,11 +43,9 @@ var availableVersions = []unversioned.GroupVersion{v1beta1.SchemeGroupVersion}
 
 func init() {
```
Comment on availableVersions (line 42) and not this line:
The ordering comment there is invalid now right?

caesarxuchao replied:
It is still valid. @deads2k has another PR related to this, but IIRC that one also relies on the `availableVersions` to be in the correct order.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
+		return
+	}
+
+	addVersionsToScheme(externalVersions...)
```
line 64 to line 82 here is the code that really "enables" the versions. I guess we can move that into an enableVersions() method and call it here.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -105,3 +112,19 @@ func interfacesFor(version unversioned.GroupVersion) (*meta.VersionInterfaces, e
+
+func addVersionsToScheme(externalVersions ...unversioned.GroupVersion) {
+	// add the internal version to Scheme
+	extensions.Register()
```
In the new terminology, it was "registered" already, right?
It is being "enabled" now?

caesarxuchao replied:
Done

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -43,11 +43,9 @@ var availableVersions = []unversioned.GroupVersion{v1beta1.SchemeGroupVersion}
 
 func init() {
 	externalVersions := []unversioned.GroupVersion{}
```
Based on all the comments below, the high level flow here should be:

func init() {
  registered.RegisterVersion()
  if (shouldEnableVersion()) {
     enableVersion()
     registered.AddEnabledVersion()
  }
}
This should make the code easier to understand.

nikhiljindal commented:
So we really have 3 levels now: registered, enabled, and exposed (whether the groupversion is exposed in the API by master, controlled by runtime-config).

Have added a few comments, found when I was trying to understand the code.

nikhiljindal commented on pkg/apis/extensions/install/install.go:
```diff
@@ -56,6 +54,15 @@ func init() {
+
 	preferredExternalVersion := externalVersions[0]
 
 	groupMeta := latest.GroupMeta{
```
As a separate issue (out of scope of this PR, I believe), all this code to enable the version should not be present here.
This code should belong with api machinery code and not with specific api code.
Someone writing a package for a new API group version should just call some genericApiMachinery.EnableVersion() and pass it the required params. the generic api machinery code should know that it needs to call latest.RegisterGroup and api.RegisterRESTMapper, not each api group version package.

caesarxuchao replied:
Added a Todo.

caesarxuchao commented on pkg/api/install/install.go:
```diff
@@ -42,20 +42,32 @@ var accessor = meta.NewAccessor()
 		glog.V(4).Infof("No version is registered for group %v", api.GroupName)
 		return
 	}
+	if err := registered.EnableVersions(externalVersions...); err != nil {
```
@nikhiljindal, per our offline discussion, the order is still marking those versions as enabled first, and then actually enable the versions, because we want to check we only add the enabled versions to scheme.

caesarxuchao commented:
@nikhiljindal PTAL. Thanks!

caesarxuchao commented on cmd/libs/go2idl/client-gen/testdata/apis/testgroup/register.go:
```diff
@@ -23,8 +23,8 @@ import (
 
-func init() {
-	// Register the API.
+func AddToScheme() {
```
@nikhiljindal I rename the `Reigster()` to `AddToScheme`.

nikhiljindal commented:
Thanks for replying to my comments. The changes LGTM.

caesarxuchao commented:
Thanks. I'll squash and apply the label per previous comments.

Commit:
pass runtime.Scheme as parameter to AddToScheme

caesarxuchao commented:
Rebased and added a quite mechanical change to address https://github.com/caesarxuchao/kubernetes/commit/43b60c9f4d995d117dfed1c51f1eadcf4bccd2c6#commitcomment-15144868.  Adding back the LGTM.

caesarxuchao commented:
@k8s-bot unit test this
